src/Grpc.AspNetCore.Server/Internal/PercentEncodingHelpers.cs

```diff
@@ -0,0 +1,218 @@
+                    if (current > AsciiMaxValue)
+                    {
+                        // Get additional unicode characters
+                        var unicodeCharCount = GetUnicodeCharCount(s, i);
+
+                        var numberOfBytes = Encoding.UTF8.GetBytes(s.AsSpan(i, unicodeCharCount), unicodeBytesBuffer);
+
+                        for (var count = 0; count < numberOfBytes; count++)
+                        {
+                            EscapeAsciiChar(ref span, ref writePosition, (char)unicodeBytesBuffer[count]);
+                        }
+                        i += unicodeCharCount - 1;
+                    }
+                    else if (IsUnreservedCharacter(current))
+                    {
+                        span[writePosition++] = current;
+                    }
+                    else
+                    {
+                        EscapeAsciiChar(ref span, ref writePosition, current);
+                    }
+                }
+            }
+        }
+
+        private static void EscapeAsciiChar(ref Span<char> span, ref int writePosition, char current)
+        {
+            span[writePosition++] = '%';
+            span[writePosition++] = HexChars[current >> 4];
+            span[writePosition++] = HexChars[current & 15];
+        }
+
+        private static int GetUnicodeCharCount(string value, int currentIndex)
+        {
+            var unicodeCharCount = 0;
+            var maxSize = Math.Min(value.Length - currentIndex, MaxUnicodeCharsReallocate - 1); // Leave a character for possible low surrogate
+            for (; unicodeCharCount < maxSize && value[currentIndex + unicodeCharCount] > AsciiMaxValue; unicodeCharCount++)
+            {
+            }
+
+            if (char.IsHighSurrogate(value[currentIndex + unicodeCharCount - 1]))
+            {
+                if (unicodeCharCount < value.Length - currentIndex && char.IsLowSurrogate(value[currentIndex + unicodeCharCount]))
+                {
+                    // Last character is a high surrogate so check ahead to see if it is followed by a low surrogate and include
+                    unicodeCharCount++;
+                }
+            }
+
+            return unicodeCharCount;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsUnreservedCharacter(char c)
+        {
+            return ((PercentEncodingUnreservedBitField[c / 8] >> (c % 8)) & 1) != 0;
+        }
+
+        // TODO(JamesNK): Remove WIP decode once we are certain Uri.UnescapeData works correctly
+#if Decode
+        public static string PercentDecode(string value)
+        {
+            var writeLength = 0;
+            var readIndex = 0;
+            while (readIndex < value.Length)
+            {
+                if (value[readIndex] == '%')
+                {
+                    if (readIndex + 2 < value.Length && IsValidHex(value[readIndex + 1]) && IsValidHex(value[readIndex + 2]))
```

Comment:
Possible integer overflow: should instead read `if (readIndex < value.Length - 2 && ...)` or `if ((uint)(readIndex + 2) < (uint)value.Length && ...)`.